### ISSUE TITLE
New Rule: disallow assigning class static field

### ISSUE THREAD

**Issue by sosukesuzuki** [eslint/eslint#16921]
### Rule details

Disallow assigning class static fields

### Related ECMAScript feature

https://github.com/tc39/proposal-class-fields

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
// incorrect
class Foo {}
Foo.field = 3;

// correct
class Foo {
  static field = 3;
}
```

### Why should this rule be in the core instead of a plugin?

This is because class static features are new in ECMAScript 2022 and this is a potential best practice.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

I am not sure if this should be included in the ESLint core. but, I think this is the same kind of rule as [no-class-assign](https://eslint.org/docs/latest/rules/no-class-assign), etc.

If this is accepted, I think the name of this rule would be something like `no-class-static-assign`.

**Comment by mdjermanovic** [comment]
Is the proposal to disallow reassignments, or to suggest defining static fields instead of assigning to properties after the class is defined? In the latter case, I think a name like `prefer-class-static-fields` would be better.

**Comment by sosukesuzuki** [comment]
`prefer-class-static-fields` sounds good to me

**Comment by sosukesuzuki** [comment]
Is this proposal acceptable?

**Comment by Rec0iL99** [comment]
> Is the proposal to disallow reassignments, or to suggest defining static fields instead of assigning to properties after the class is defined?

@sosukesuzuki there's a query for you to answer. Doing so we can move forward after approval.

**Comment by sosukesuzuki** [comment]
I thought I answered the question at https://github.com/eslint/eslint/issues/16921#issuecomment-1441722452, sorry.
This proposed rule suggests defining static fields instead of assigning to properties after the class is defined.

**Comment by nzakas** [comment]
I'm :-1: to this proposal as it is just enforcing a somewhat obscure preference and not preventing an error. I can see why you might want such a rule, but I think it's better off as a plugin rule rather than a core rule.

**Comment by sam3k** [comment]
We've decided in the 2023-03-23 TSC meeting to not accept this proposal. Notes from this meeting will be published in this [PR](https://github.com/eslint/tsc-meetings/pull/419/files).

**Comment by sosukesuzuki** [comment]
Thank you team for discuss about this proposal!
